src/layers/attention.h

```diff
@@ -755,60 +771,74 @@ class Attention {
                 }
             }
         }
-        free(transQKV);
     }
 
     // scaled dot-product attention: bmm1 + softmax + bmm2
-    void scaledDpAttention(const float *query, const float *key, const float *value, const float *attnMask, float scale,
-            int batchSize, int srcLen, int tgtLen, int numQHead, int numKVHead, int headSize, float *output) {
+    template <typename AttnT>
+    void scaledDpAttention(const float *query, const AttnT *key, const AttnT *value, const float *attnMask,
+            float scale, int batchSize, int srcLen, int tgtLen, int numQHead, int numKVHead, int headSize,
+            float *output, int qStride, int kvStride, int stride) {
         // output = trans(softmax(query * trans(key)) * value)
         int nth = omp_get_max_threads();
-        int minBlk = (nth >= batchSize * numQHead ? 256 : 512);
-        int srcBlk = std::min(minBlk, srcLen);
-        int tgtBlk = std::min(minBlk, tgtLen);
+        int minBlk = (int)std::pow(2, int(std::log2(srcLen / 2)));
```

Comment:
any design principle here? if any, would you make some comment?

Reply:
The current block size is just derived from practical experience. Added comments.